python/cusignal/windows/windows.py

```diff
@@ -1827,6 +1827,138 @@ def exponential(M, center=None, tau=1.0, sym=True):
+        are less than or equal to 1. When False the DC gain will remain at 1
+        (0 dB) and the sidelobes will be `sll` dB down.
+    sym : bool, optional
+        When True (default), generates a symmetric window, for use in filter
+        design.
+        When False, generates a periodic window, for use in spectral analysis.
+    Returns
+    -------
+    out : array
+        The window. When `norm` is True (default), the maximum value is
+        normalized to 1 (though the value 1 does not appear if `M` is
+        even and `sym` is True).
+    See Also
+    --------
+    chebwin, kaiser, bartlett, blackman, hamming, hanning
+    References
+    ----------
+    .. [1] W. Carrara, R. Goodman, and R. Majewski, "Spotlight Synthetic
+           Aperture Radar: Signal Processing Algorithms" Pages 512-513,
+           July 1995.
+    .. [2] Armin Doerry, "Catalog of Window Taper Functions for
+           Sidelobe Control", 2017.
+           https://www.researchgate.net/profile/Armin_Doerry/publication/316281181_Catalog_of_Window_Taper_Functions_for_Sidelobe_Control/links/58f92cb2a6fdccb121c9d54d/Catalog-of-Window-Taper-Functions-for-Sidelobe-Control.pdf
+    Examples
+    --------
+    Plot the window and its frequency response:
+    >>> from scipy import signal
+    >>> from scipy.fft import fft, fftshift
+    >>> import matplotlib.pyplot as plt
+    >>> window = signal.windows.taylor(51, nbar=20, sll=100, norm=False)
+    >>> plt.plot(window)
+    >>> plt.title("Taylor window (100 dB)")
+    >>> plt.ylabel("Amplitude")
+    >>> plt.xlabel("Sample")
+    >>> plt.figure()
+    >>> A = fft(window, 2048) / (len(window)/2.0)
+    >>> freq = np.linspace(-0.5, 0.5, len(A))
+    >>> response = 20 * np.log10(np.abs(fftshift(A / abs(A).max())))
+    >>> plt.plot(freq, response)
+    >>> plt.axis([-0.5, 0.5, -120, 0])
+    >>> plt.title("Frequency response of the Taylor window (100 dB)")
+    >>> plt.ylabel("Normalized magnitude [dB]")
+    >>> plt.xlabel("Normalized frequency [cycles per sample]")
+    """  # noqa: E501
+    if _len_guards(M):
+        return np.ones(M)
+    M, needs_trunc = _extend(M, sym)
+
+    # Original text uses a negative sidelobe level parameter and then negates
+    # it in the calculation of B. To keep consistent with other methods we
+    # assume the sidelobe level parameter to be positive.
+    B = 10 ** (sll / 20)
+    A = np.arccosh(B) / np.pi
+    s2 = nbar ** 2 / (A ** 2 + (nbar - 0.5) ** 2)
+    ma = np.arange(1, nbar)
+
+    Fm = np.empty(nbar - 1)
+    signs = np.empty_like(ma)
+    signs[::2] = 1
+    signs[1::2] = -1
+    m2 = ma * ma
+    for mi, _ in enumerate(ma):
+        numer = signs[mi] * np.prod(
+            1 - m2[mi] / s2 / (A ** 2 + (ma - 0.5) ** 2)
+        )
+        denom = (
+            2
+            * np.prod(1 - m2[mi] / m2[:mi])
+            * np.prod(1 - m2[mi] / m2[mi + 1 :])
+        )
+        Fm[mi] = numer / denom
```

Comment:
Are these parameter settings/small array manipulations that are faster in numpy than cupy?

Reply:
`Fm` is an array of size == nbar, which should be much smaller than M.

The `Fm` array is used by each thread computing a value in `w`.

It is much easier to compute on CPU and copy `Fm` to global memory. Depending on the size it should reside in L2, and maybe L1. It would require a very large `nbar` to justify moving to workload to GPU. It would also require a more complex custom CUDA kernel with shared memory usage.

Reply:
Thanks for the clarification. LGTM!

Reply:
I would be interesting to discuss the idea, with @gmarkall, of moving small for loops like this to Numba CPU.

Reply:
But even then, it's probably only justified if the function is ran multiple times...

Reply:
Yeah, the typical workflow here is to create a window function and then successively apply it to new data. The filtering operation is performed over and over, but not the window creation.

Reply:
Is the function constant across runs? The CPU target has on-disk caching, so if the compilation overhead is a problem here then caching could help reduce to overhead to justify using Numba for it.